src/infer_payload.cc

```diff
@@ -38,13 +38,15 @@ InferPayload::InferPayload(
 
 InferPayload::~InferPayload()
 {
+  std::lock_guard<std::mutex> lock(mutex_);
   prev_promise_.reset();
 }
 
 void
 InferPayload::SetValueForPrevPromise(
     std::unique_ptr<InferResponse> infer_response)
 {
+  std::lock_guard<std::mutex> lock(mutex_);
```

Comment:
Question: I think we should bring in the full semantics from request_executor.cc in order to have well defined behavior:
Otherwise (if I understand correctly) - there is a possibility that two responses could both see the promise as not set and then one could be dropped? We could update the method name to be clearer: "SetValue" 

```
std::lock_guard<std::mutex> lock(mutex_);
if (! is_promise_set_) {
     is_promise_set_= true;
     prev_promise_->set_value(std::move(infer_response));
} else {
    Callback(std::move(infer_response));
}

```